graph/multigraph.go

```diff
@@ -125,10 +125,10 @@ type WeightedLineAdder interface {
 
 // LineRemover is an interface for removing lines from a multigraph.
 type LineRemover interface {
-	// RemoveLine removes the given line, leaving the
-	// terminal nodes. If the line does not exist it
-	// is a no-op.
-	RemoveLine(Line)
+	// RemoveLine removes the line with the given end
+	// and line IDs, leaving the terminal nodes. If
+	// the line does not exist it is a no-op.
+	RemoveLine(fid, tid, id int64)
```

Comment:
Something about the `RemoveLine` method doesn't feel right to me. It is too many IDs to keep track of, and not entirely clear what each one is referring to.

I know, that `fid` is from ID, `tid` is to ID and `id` is the line ID. but this is clear only after a moments thought. Also, the line `id` is not analogous to the node ID which increases the cognitive load.

Perhaps this would be a good time to discuss whether the line ID should be unique in the entire graph, rather than just for edges with the same end points in a multigraph. What would the pros and cons be?

Pro:
* reduced cognitive load as `line.ID()` would by analogous to `node.ID()`.
* the `RemoveLine` method could be simplified to `RemoveLine(id int64)`

Cons:
* imposes constraints on implementations to keep track of the line IDs for the entire graph.

I would be curious to learn more about various implementations, is the con true for real world code? Would the implementation be much simpler (or efficient) of the ID only needs to be kept distinct for lines sharing the same end point?

On an extended note (one I don't know if it adds any benefit, but definitely worth discussing, and potentially rejecting). If after discussion we reach consensus that the line ID should be unique in the entire graph and not just for lines sharing the same end points, then it would be reasonable to open the question of whether we can to represent edges with an edge ID?

If that would be the case than the entire distinction between edges and lines could be removed.

I understand this opens a larger question than perhaps this PR intended. However, it is better to get the API right now, rather than at some later point in the future, as both this PR and the suggested change to make line IDs unique in the graph cause API breakage.

What are your thoughts on this?

Reply:
>I know, that `fid` is from ID, `tid` is to ID and `id` is the line ID. but this is clear only after a moments thought. Also, the line `id` is not analogous to the node ID which increases the cognitive load.

I disagree here; the line ID is exactly analogous to a node ID, though in a different namespace.

>Perhaps this would be a good time to discuss whether the line ID should be unique in the entire graph, rather than just for edges with the same end points in a multigraph. What would the pros and cons be?

Ah, OK, I see what you mean by not analogous. If edges had an ID then all edge removals would be by edge ID.

>Pro:
>* reduced cognitive load as `line.ID()` would by analogous to `node.ID()`.
>* the `RemoveLine` method could be simplified to `RemoveLine(id int64)`

>Cons:
>* imposes constraints on implementations to keep track of the line IDs for the entire graph.

Note that this is actually how the graphs implement them, though that is not currently required.

**EDIT:** and of course, point out more pros and cons if you think of any.

>I would be curious to learn more about various implementations, is the con true for real world code? Would the implementation be much simpler (or efficient) of the ID only needs to be kept distinct for lines sharing the same end point?

In some cases it would be a lot simpler to not bother with retaining a graph-wide ID list; the most obvious case is when lines are stored in a `[]graph.Line` and the ID is just the slice index - this is not currently prohibited (i.e. we don't say that the line ID must persist between calls).

>On an extended note (one I don't know if it adds any benefit, but definitely worth discussing, and potentially rejecting). If after discussion we reach consensus that the line ID should be unique in the entire graph and not just for lines sharing the same end points, then it would be reasonable to open the question of whether we can also represent edges with an edge ID?

>If that would be the case than the entire distinction between edges and lines could be removed.

This is actually how graph was for a brief period after #335 went in. I prefer the simpler computational, cognitive and SLOC burden on the common simple case, I'm also reasonably happy with the separation between edges and lines and how they are reintegrated as a `[]graph.Line` being able to be a `graph.Edge`.

>I understand this opens a larger question than perhaps this PR intended. However, it is better to get the API right now, rather than at some later point in the future, as both this PR and the suggested change to make line IDs unique in the graph cause API breakage.

Agreed.

Overall, I'm not sure about what to do with this. @vladimir-ch?

Reply:
Can we move forward with this?

Reply:
I agree that the signature of `RemoveLine` is not very nice, I also had pause to parse the parameters and think which is which. If lines had graph-unique IDs, then edges should have IDs, then it would not make sense/be necessary to distinguish between lines and edges. On the downside, without `Lines` there would be no clear distinguishing method for `Multigraph`. The question is if `Multigraph` distinction would still be needed. So the consequences are big and it's not clear that merging Edges and Lines and giving them IDs were beneficial, so I'd leave it as is for the moment.

Reply:
The signature here very closely matches the mathematics though (in both the simple and multigraph cases).

Reply:
Yes, let's keep this signature.